OpenTreeMap/otm-core
# Cache species and boundary autocomplete data

See https://github.com/twitter/typeahead.js/blob/master/doc/bloodhound.md
for a description of cache parameters discussed here.

We were re-fetching an instance's species and boundary data on every page load
because we were using `ttl` ("time to live") of 0.
Fix is to use a default `ttl` of one day, and a `thumbprint` (cache buster) based on
the latest update time of the appropriate table (Species or Boundary).

Also, visiting too many instances was causing a "quota exceeded" error for browser local storage.
Fix is to use an instance-independent `cacheKey` ("species" or "boundary").
That means switching instances invalidates the cache, but that's rare compared to
visiting many consecutive pages on the same instance.

Connects #1857
Connects #2019
Connects #199

To test:

0) Clear your browser local storage
* window.localStorage.clear()
* window.localStorage

1) Verify that species and boundary autocomplete data is cached
* Open developer tools, network tab
* Load a treemap
* Note that `/instance/species/` and `/instance/boundaries/` are fetched.
  (The species are fetched twice. Hard to avoid.)
* Go to a plot detail page. Note that species and boundaries are not re-fetched.

2) Verify that loading a different treemap invalidates the cache
* Load a different treemap
* Note that the search dropdowns contain different species and boundaries

3) Verify that changing data invalidates the appropriate cache
* Start a Django shell
* `from treemap.models import Species, Boundary`
* `s = Species.objects.get(instance__url_name='latreemap', common_name='Afghan pine')`
* `s.common_name='Afghan pony'`
* `s.save_with_system_user_bypass_auth()`
* `s = Boundary.objects.get(name='Alhambra')`
* `s.name = 'Alhamburger'`
* `s.save()`
* Reload the appropriate treemap
* See your changes in the dropdowns
